MovingBlocks/Terasology
# Add facet types and spawn logic for sparsely-valued spawn heights and sea level heights.

Closes #3177

### Contains

Adds two new facet types, that allow spawn heights and sea level heights to not be defined for all XZ coordinates. If a spawn height is not defined for a coordinate, the player cannot spawn there. If a sea level height is not defined for a coordinate, the player can spawn there, regardless of height.

If there is no spawn height for which the player is above sea level, the player will be spawned without regard to sea level. It's still possible for there to not be a valid spawn location (e.g. no spawn heights specified), in which case this is considered a world generation error, and an exception is thrown.

### How to test

- Ensure that this PR remains backwards compatible with other world generators, which should remain unaffected in terms of spawn logic

- Use the new facets, and ensure that spawn locations are being chosen as expected
